### PR TITLE
Face-2-Colouring

### PR DESCRIPTION
**Pull request by ReymondAkpanya**
#106

### CODE REVIEW

File: gap/PolygonalComplexes/properties.gi [hunk 1]
```diff
@@ -918,4 +918,44 @@ InstallMethod( IsQuadrangular,
...
+##      Face colouring
+##
+
+InstallMethod( FaceTwoColouring, 
+    "for a polygonal complex",
+	[IsPolygonalComplex],
+	function(complex)
+		local faces,red,blue,f,remfaces,tempfaces,neighbours;
+		if 1 in List(FaceDegreesOfVertices(complex),g->g mod 2) then 
+			Error("Two Colouring does not exist");
```
Comment (MeikeWeiss) [inline gap/PolygonalComplexes/properties.gi:932]: It would be better to return fail and not an error.

File: gap/PolygonalComplexes/properties.gi [hunk 2]
```diff
@@ -918,4 +918,44 @@ InstallMethod( IsQuadrangular,
...
+		fi;
+		blue:=[Faces(complex)[1]];
+		red:=[];
+		remfaces:=Difference(Faces(complex),blue);
+		while remfaces <> [] do 
+			tempfaces:=Difference(Faces(complex),Union(blue,red));
```
Comment (MeikeWeiss) [inline gap/PolygonalComplexes/properties.gi:938]: At this place tempfaces and remfaces are the same, so this line is unnecessary if you use in the next line remfaces in the filter.

File: gap/PolygonalComplexes/properties.gi [hunk 3]
```diff
@@ -918,4 +918,44 @@ InstallMethod( IsQuadrangular,
...
+			tempfaces:=Filtered(tempfaces,f->Intersection(NeighbourFacesOfFace(complex,f),Union(blue,red))<>[]);
+			for f in tempfaces do 
+				neighbours:=NeighbourFacesOfFace(complex,f);
+				if Intersection(neighbours,blue)=[] then
+					Add(blue,f);
+				elif Intersection(neighbours,red)=[] then 
+					Add(red,f);
+				else
+					Error("Two colouring does not exist");
```
Comment (MeikeWeiss) [inline gap/PolygonalComplexes/properties.gi:947]: Here it would be also better to return fail and not an error.

**Review by MeikeWeiss — CHANGES_REQUESTED** [review]
This looks good so far. Some improvement can be made.

**Review by markusbaumeister — CHANGES_REQUESTED** [review]
This looks good so far.
There are some small issues that should be adressed before merging.

**Comment by markusbaumeister** [comment]
I just checked the manual as well: can you put the method into its own subsection? You just need to insert the following snippet in front of its documentation in properties.gd:
`#! @Section Face-colouring
#! @SectionLabel Properties_FaceColouring
`

File: gap/PolygonalComplexes/properties.gd [hunk 4]
```diff
@@ -763,3 +763,25 @@ DeclareOperation( "IsChaoticVertex", [IsTwistedPolygonalComplex, IsPosInt] );
 DeclareOperation( "IsChaoticVertexNC", [IsTwistedPolygonalComplex, IsPosInt] );
 #! @EndGroup
 
+#! @BeginGroup FaceTwoColouring
+#! @Description
+#! The method returns a 2-face-colouring of a given polygonal complex if it exists
+#! or an error if it doesn't.
```
[the remark this answers is not in the document]
Comment (markusbaumeister) [reply gap/PolygonalComplexes/properties.gd]: Apparently, GitHub did not show my full feedback here. It would be nice if "fail" was written as `<K>fail</K>` in the documentation, since this ensures a slightly different formatting.

File: gap/PolygonalComplexes/properties.gd [hunk 5]
```diff
@@ -763,3 +763,25 @@ DeclareOperation( "IsChaoticVertex", [IsTwistedPolygonalComplex, IsPosInt] );
...
+#! @ExampleSession
+#! gap> FaceTwoColouring(Octahedron());
+#! [ [ 1, 2, 5, 6 ], [ 3, 4, 7, 8 ] ]
+#! @EndExampleSession
+#! <Alt Only="TikZ">
+#!      \input{Image_TwoColouredOctahedron.tex}
+#! </Alt>
+#! @Returns a pair where the first and second entry are sets of faces or an Error
+#! @Arguments complex
+DeclareAttribute( "FaceTwoColouring", IsPolygonalComplex );
```
[the remark this answers is not in the document]
Comment (markusbaumeister) [reply gap/PolygonalComplexes/properties.gd:788]: This seems not adressed so far.
By the way, this is also an issue for polygonal complexes that are not strongly connected.

File: gap/PolygonalComplexes/properties.gi [hunk 6]
```diff
@@ -918,4 +918,44 @@ InstallMethod( IsQuadrangular,
...
+##
+
+InstallMethod( FaceTwoColouring, 
+    "for a polygonal complex",
+	[IsPolygonalComplex],
+	function(complex)
+		local faces,red,blue,f,remfaces,tempfaces,neighbours;
+		if 1 in List(FaceDegreesOfVertices(complex),g->g mod 2) then 
+			Error("Two Colouring does not exist");
+		fi;
```
[the remark this answers is not in the document]
Comment (markusbaumeister) [reply gap/PolygonalComplexes/properties.gi]: This is not adressed so far.

File: gap/PolygonalComplexes/properties.gd (hunk 5 above)
[the remark this answers is not in the document]
Comment (ReymondAkpanya) [reply gap/PolygonalComplexes/properties.gd:788]: Since Faces(comp) is ordered and I start by assigning  the color blue to the first face (face with the smallest label), the uniquness is quaranteed.

[the remark this answers is not in the document]
Comment (markusbaumeister) [reply gap/PolygonalComplexes/properties.gd:788]: Sure, but you should mention that in the method documentation :)

**Review by markusbaumeister — CHANGES_REQUESTED** [review]
There have been some minor improvements, but several critical aspects remain unaddressed. Concretely:
-> The current code gives incorrect results for all polygonal complexes that are not closed surfaces.
-> The current code cannot handle disconnected complexes.
Please add some tests for cases like this (disconnected complexes, surfaces with boundary, complexes with ramifications). These should probably go into unit_tests/ to avoid cluttering the manual.

**Review by markusbaumeister — APPROVED** [review]
This looks ok now.
